File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/unpacker/ArgmaxUnpacker.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing.unpacker;

import com.facebook.react.bridge.WritableMap;
import org.json.JSONException;
import org.json.JSONObject;
import org.pytorch.IValue;
import org.pytorch.rn.core.ml.processing.PackerContext;
import org.pytorch.rn.core.ml.processing.Unpacker;

public class ArgmaxUnpacker implements Unpacker {

  private static final String VALUE_KEY = "valueKey";

  @Override
  public void unpack(
      IValue ivalue, JSONObject jobject, WritableMap result, PackerContext packerContext)
      throws JSONException {
    final String key = jobject.getString("key");
    String valueKey = null;
    if (jobject.has(VALUE_KEY)) {
      valueKey = jobject.getString(VALUE_KEY);
    }
    final String dtype = jobject.getString("dtype");

    int maxIdx;
    switch (dtype) {
      case "float":
        float[] floatData = ivalue.toTensor().getDataAsFloatArray();
        maxIdx = unpackArgmaxFloatData(floatData);
        result.putInt(key, maxIdx);
        if (valueKey != null) {
          float[] floatSoftmax = softmax(floatData);
          result.putDouble(valueKey, floatSoftmax[maxIdx]);
        }
        break;
      case "long":
        long[] longData = ivalue.toTensor().getDataAsLongArray();
        maxIdx = unpackArgmaxLongData(longData);
        result.putInt(key, maxIdx);
        if (valueKey != null) {
          float[] floatSoftmax = softmax(longData);
          result.putDouble(valueKey, floatSoftmax[maxIdx]);
        }
        break;
      default:
        throw new IllegalArgumentException("Unknown dtype");
    }
  }

  private static int unpackArgmaxFloatData(float[] data) {
    float maxValue = -Float.MAX_VALUE;
    int maxIdx = -1;
    for (int i = 0; i < data.length; i++) {
      if (data[i] > maxValue) {
        maxValue = data[i];
        maxIdx = i;
      }
    }
    return maxIdx;
  }

  private static int unpackArgmaxLongData(long[] data) {
    long maxValue = -Long.MAX_VALUE;
    int maxIdx = -1;
    for (int i = 0; i < data.length; i++) {
      if (data[i] > maxValue) {
        maxValue = data[i];
        maxIdx = i;
      }
    }
    return maxIdx;
  }

  private static float[] softmax(float[] data) {
    final float[] result = new float[data.length];
    float expSum = 0;

    for (int i = 0; i < data.length; i++) {
      result[i] = (float) Math.exp(data[i]);
      expSum += result[i];
    }

    for (int i = 0; i < data.length; i++) {
      result[i] /= expSum;
    }
    return result;
  }

  private static float[] softmax(long[] data) {
    final float[] result = new float[data.length];
    float expSum = 0;

    for (int i = 0; i < data.length; i++) {
      result[i] = (float) Math.exp(data[i]);
      expSum += result[i];
    }

    for (int i = 0; i < result.length; i++) {
      result[i] /= expSum;
    }
    return result;
  }
}

File: react-native-pytorch-core/ios/ML/Processing/Unpacker/ArgmaxUnpacker.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

class ArgmaxUnpacker: Unpacker {
    func unpack(ivalue: IValue, modelSpec: JSON, result: inout [String: Any], packerContext: PackerContext) throws {
        let unpack = modelSpec["unpack"]

        let dtype = unpack["dtype"].string
        guard let key = unpack["key"].string else {
            throw BaseIValuePackerError.missingKeyParam
        }
        let valueKey = unpack["valueKey"].string
        switch dtype {
        case "float":
            let tensor = ivalue.toTensor()
            guard let array = tensor?.getDataAsArray() else {
                throw BaseIValuePackerError.invalidParam
            }
            let maxIdx = argmax(array: array, dtype: tensor?.dtype)
            result[key] = maxIdx
            if valueKey != nil {
                let softmaxData = softmax(data: array)
                result[valueKey!] = softmaxData[maxIdx]
            }
        default:
            throw BaseIValuePackerError.invalidDType
        }
    }

    func softmax(data: [NSNumber]) -> [Double] {
        return softmax(data: data, fromIndex: 0, toIndex: data.count)
    }

    func softmax(data: [NSNumber], fromIndex: Int, toIndex: Int) -> [Double] {
        var softmax = [Double](repeating: 0.0, count: (toIndex - fromIndex))
        var expSum = 0.0

        for idx in fromIndex..<toIndex {
            let expValue = exp(data[idx].doubleValue)
            softmax[idx - fromIndex] = expValue
            expSum += expValue
        }

        for idx in 0..<softmax.count {
            softmax[idx] /= expSum
        }

        return softmax
    }

    private func argmax(array: [Any]?, dtype: PTMTensorType?) -> Int {
        guard let array = array, let dtype = dtype else {
            return 0
        }

        switch dtype {
        case .float:
            var max = -MAXFLOAT
            var ret = -1
            for (idx, item) in array.enumerated() {
                guard let num = item as? Float32 else {
                    continue
                }
                if num > max {
                  ret = idx
                  max = num
                }
            }
            return ret
        default:
            return 0
        }
    }
}
